deterministic_kernel.hpp

```diff
@@ -0,0 +1,240 @@
+    CauchyDistribution cauchy;
+    ExponentialDistribution exponential;
+    double number_of_dispersers;
+public:
+    DeterministicDispersalKernel(DispersalKernelType dispersal_kernel,
+                                const IntegerRaster& dispersers, double dispersal_percentage,
+                                double ew_res, double ns_res, double distance_scale = 1.0,
+                                double locator = 0.0)
+        :
+        dispersers_(dispersers),
+        cauchy(distance_scale, locator),
+        exponential(distance_scale),
+        kernel_type_(dispersal_kernel),
+        east_west_resolution(ew_res),
+        north_south_resolution(ns_res)
+    {
+        if (kernel_type_ == DispersalKernelType::Cauchy) {
+            max_distance = cauchy.icdf(dispersal_percentage);
+        } else if (kernel_type_ == DispersalKernelType::Exponential) {
+            max_distance = exponential.icdf(dispersal_percentage);
+        } else {
+            throw std::invalid_argument("Unsupported dispersal kernel type");
+        }
+        number_of_columns = ceil(max_distance / east_west_resolution) * 2 + 1;
+        number_of_rows = ceil(max_distance / north_south_resolution) * 2 + 1;
+        Raster<double> prob_size(number_of_rows, number_of_columns, 0);
+        probability = prob_size;
+        probability_copy = prob_size;
+        mid_row = number_of_rows / 2;
+        mid_col = number_of_columns / 2;
+        double sum = 0.0;
+        for ( int i = 0; i < number_of_rows; i++ ) {
+            for ( int j = 0; j < number_of_columns; j++ ) {
+                double distance_to_center =
+                std::sqrt(pow((abs(mid_row - i) * east_west_resolution), 2)
+                + pow((abs(mid_col - j) * north_south_resolution), 2));
+                // determine probability based on distance
+                if (kernel_type_ == DispersalKernelType::Cauchy) {
+                    probability(i,j) = abs(cauchy.pdf(distance_to_center));
+                } else if ( kernel_type_ == DispersalKernelType::Exponential ) {
+                    probability(i,j) = abs(exponential.pdf(distance_to_center));
+                }
+            sum += probability(i,j);
+            }
+        }
+        //normalize based on the sum of all probabilities in the raster
+        probability /= sum;
+    }
+
+    /*! Generates a new position for the spread.
+     *
+     *  Creates a copy of the probability matrix to mark where dispersers are assigned.
+     *  New window created any time a new cell is selected from simulation.disperse
+     *
+     *  Selects next row/col value based on the cell with the highest probability
+     *  in the window.
+     *
+     */
+    template<class Generator>
+    std::tuple<int, int> operator()(Generator& generator, int row, int col)
+    {
+        // reset the window if considering a new cell
+        if ( row != prev_row || col != prev_col ) {
+            number_of_dispersers = 1.0 / (double)dispersers_(row, col);
```

Comment:
rename number_of_dispersers to proportion_of_dispersers to be more precise in the naming and easier to follow